gflownet/envs/tetris.py

```diff
@@ -138,30 +152,73 @@ def _drop_piece_on_board(
         board = state.clone().detach()
 
         piece_idx, rotation, col = action
-        piece_mat = torch.rot90(
-            self.piece2mat(self.idx2piece[piece_idx]), k=self.rot2idx[rotation]
-        )
+        piece_mat = self.piece_rotation_mat[self.idx2piece[piece_idx]][rotation]
+        piece_mat_mask = self.piece_rotation_mask_mat[self.idx2piece[piece_idx]][
+            rotation
+        ]
         hp, wp = piece_mat.shape
-        # Get and set index of new piece
-        indices = board.unique()
-        piece_idx = self._get_max_piece_idx(board, piece_idx, incr=1)
-        piece_mat[piece_mat != 0] = piece_idx
+
         # Check if piece goes overboard horizontally
         if col + wp > self.width:
             return board, False
-        # Iteratively attempts to place piece on the board starting from the bottom
-        for row in reversed(range(self.height)):
-            if row - hp + 1 < 0:
-                return board, False
-            board_section = board[row - hp + 1 : row + 1, col : col + wp]
-            # If all board cells under piece are empty
-            if sum(board_section[piece_mat != 0]) == 0:
-                board_aux = board.clone().detach()
-                board_aux[row - hp + 1 : row + 1, col : col + wp] += piece_mat
-                # If all columns above piece are empty
-                if self._piece_can_be_lifted(board_aux, piece_idx):
-                    return board_aux, True
-        return board, False
+
+        # Find the highest row occupied by any other piece in the columns where we wish
+        # to add the new piece
+        occupied_rows = board[:, col : col + wp].sum(1).nonzero()
+        if len(occupied_rows) == 0:
+            # All rows are unoccupied. Set first occupied row as the row "below" the
+            # board.
+            first_occupied_row = self.height
+        else:
+            first_occupied_row = occupied_rows[0, 0]
```

Comment:
"first" seems a bit ambiguous, no?
```suggestion
            highest_occupied_row = occupied_rows[0, 0]
```